### ISSUE TITLE
Return input object if change does not result in an op

### ISSUE THREAD

**Issue by mattkrick** [automerge/automerge#107]
I noticed that every time `change()` is called an item gets added to the history, even if no action was performed.

```js
const a = Automerge.init({})
const b = Automerge.change(a, () => {})

b._state.get('opSet').get('history').size // 1
a === b // false
```

Sometimes, I have to look at an object to know if I want to change it.
I can do this just fine today by doing all my logic before calling `change`, but that means looking up things once before calling change, and then again once inside.

I think it'd be great to instead just return the input value if nothing changed.
Then, folks could do something like

```js
nextValue = Automerge.change(val, () => {})
if (nextValue !== val) broadcastChanges(nextValue, val)
```

**Comment by kpruden** [comment]
I has a similar question, but I asked in slack: https://automerge.slack.com/archives/C61RJCM9S/p1531413629000204

This was the response from @ept:

> Good question. I'm a bit torn actually. On the one hand it makes sense to not pollute the history with unnecessary changes. On the other hand, if a change description was supplied (as second argument to `Automerge.change`), then silently dropping that text seems like it could be confusing to developers.
I think I'm inclined to leave the behaviour as-is right now. If you want the optimisation, you can get it by checking the most recent change object (from `getHistory()`) after making a change, and if it contains no ops, replace your Automerge document object with the one before you made the change. Since that object is immutable, you can jump back to the prior state in which the no-op change never happened.

**Comment by j-f1** [comment]
Perhaps a utility function? `doc = Automerge.revertNoops(doc)`?

**Comment by kpruden** [comment]
Ooh, that's a good idea. I've already written logic that does this, in my own code. I might be able to extract it into a PR in the next week or so..

**Comment by kpruden** [comment]
Proposed interface:
```
doc = Automerge.changeAndOmitNoops(doc, callback);
```
(I'm open to suggestions on a better/shorter name)

This would be a wrapper to `Automerge.change` and returns the originally passed `doc` if the callback produces no `ops`. The advantage of this over a separate `Automerge.revertNoops` is the caller can determine if there were any changes using simple `===`.

**Comment by ept** [comment]
You could use the following:

```js
function changeAndOmitNoops(doc, callback) {
  const updated = Automerge.change(doc, callback)
  const history = Automerge.getHistory(updated)
  if (history[history.length - 1].change.ops.length === 0) {
    return doc
  } else {
    return updated
  }
}
```

As it's such a short and simple piece of code, I think it would be better not to add it to the Automerge API, since it would create additional API surface area for fairly little benefit.

However, if the general feeling here is that omitting no-ops would be a better default behaviour, I'm open to changing my mind and making the `changeAndOmitNoops` the default. Is there anybody here who thinks that no-op changes *should* be preserved?

**Comment by kpruden** [comment]
I'm in favor of omitting no-op changes by default. Maybe it makes sense to provide a way to explicitly tell Automerge to preserve empty changes (akin to `git commit --allow-empty`). I suppose you could treat the presence of a `comment` as a signal that the caller wants the change recorded regardless, but that feels a bit magical to me.

To provide a more forward-compatible API, it might make sense to replace the `comment` parameter with an `options` parameter that can contain whatever desired modifiers to the default behavior the caller might like (including `comment`). So:
```
doc = Automerge.change(doc, 'My comment', doc => {...});
```
Becomes:
```
doc = Automerge.change(doc, { comment: 'My comment', preserveNoOp: true }, doc => {...});
```
Or something like that...

**Comment by j-f1** [comment]
Note that you can currently pass objects as commit messages.

**Comment by kpruden** [comment]
Ah, that's true. So, for backwards compatibility the signature would have to be different (maybe `options` is the last argument?), but I think the gist of my suggestion makes sense.

I'm not a native javascript developer so the convention of having optional, positional parameters not be the last argument(s) always struck me as strange :)

**Comment by mattkrick** [comment]
I'm in favor of omitting no-ops, so I'll try to plead my case :smile:
The behavior was a bit of a gotcha for me because some other popular packages omit no-ops :
- in redux, if no reducer picks up the action, the original state is returned
- in immutablejs, if you set the same key to the same value, the original object is returned
- in git commits you need a flag (as @kpruden pointed out)

My 2nd argument is that omitting no-ops opens to door to doing exactly what immutablejs does with equal value commits.
today, setting a key to the value that it's already at results in a 2nd operation. If my goal is to minimize the number of operations, I have to do a few equality checks, which start making my code look gross.

In short, I'd like to turn this:
```js
const fooChanged = val.foo !== 'bar'
const foo2Changed = val.foo2 !== 'baz'
if (fooChanged || foo2Changed) {
  Automerge.change(val, (doc) => {
    if (fooChanged) {
      doc.foo = 'bar'
    }
    if (foo2Changed) {
      doc.foo2 = 'baz'
    }
  })
}
```

into this:

```js
Automerge.change(val, (doc) => {
  doc.foo = 'bar'
  doc.foo2 = 'baz'
})
```

**Comment by jeffpeterson** [comment]
hypermerge uses no-ops to bump the vector clock after [forking](https://github.com/automerge/hypermerge/blob/master/index.js#L166-L169) and [merging](https://github.com/automerge/hypermerge/blob/master/index.js#L184-L187).

Generally, I'd say I have a slight preference for `Automerge.change` skipping empty commits, and then adding something like `Automerge.emptyChange(doc, msg)`.

**Comment by ept** [comment]
Ok, as the opinion here seems pretty much unanimous, I've made this change on PR #111. Feedback welcome!

**Comment by ept** [comment]
I have merged #111, which resolves this issue. Thank you all for contributing to this discussion!
